File: backend/src/models/Registration.ts
import mongoose, { Schema } from 'mongoose';

export type RegistrationStatus = 'registered' | 'cancelled' | 'attended';

export interface IRegistration extends mongoose.Document {
  student: mongoose.Types.ObjectId;
  event: mongoose.Types.ObjectId;
  status: RegistrationStatus;
  registeredAt: Date;
  checkedInAt?: Date;
  createdAt: Date;
  updatedAt: Date;
}

const registrationSchema = new Schema<IRegistration>(
  {
    student: { type: Schema.Types.ObjectId, ref: 'User', required: true },
    event: { type: Schema.Types.ObjectId, ref: 'Event', required: true },
    status: {
      type: String,
      enum: ['registered', 'cancelled', 'attended'],
      default: 'registered',
      required: true
    },
    registeredAt: { type: Date, default: Date.now },
    checkedInAt: { type: Date }
  },
  { timestamps: true }
);

registrationSchema.index({ student: 1, event: 1 }, { unique: true });

const Registration = mongoose.model<IRegistration>('Registration', registrationSchema);

export default Registration;

File: backend-spring/src/main/java/com/campusconnect/model/Registration.java
package com.campusconnect.model;

import lombok.Data;
import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.annotation.Id;
import org.springframework.data.annotation.LastModifiedDate;
import org.springframework.data.mongodb.core.index.CompoundIndex;
import org.springframework.data.mongodb.core.mapping.Document;

import java.time.Instant;

@Data
@Document(collection = "registrations")
@CompoundIndex(name = "student_event_idx", def = "{'student': 1, 'event': 1}", unique = true)
public class Registration {

    @Id
    private String id;

    private String student; // User id

    private String event; // Event id

    private String status = "registered"; // registered, cancelled, attended

    private Instant registeredAt = Instant.now();

    private Instant checkedInAt;

    @CreatedDate
    private Instant createdAt;

    @LastModifiedDate
    private Instant updatedAt;
}
